consensus/exit_probability_randomizer/walker.c

```diff
@@ -0,0 +1,320 @@
+    if (iter_idx++ == idx) {
+      *has_next_approver = true;
+      memcpy(approver->trits, curr_approver->hash, FLEX_TRIT_SIZE_243);
+      approver->num_bytes = FLEX_TRIT_SIZE_243;
+      break;
+    }
+  }
+
+  return RC_OK;
+}
+
+retcode_t find_tail_if_valid(const ep_randomizer_t *exit_probability_randomizer,
+                             const trit_array_t *tx_hash,
+                             trit_array_t *tail_hash, bool *found_tail) {
+  retcode_t res;
+
+  iota_stor_pack_t tx_pack;
+
+  tx_pack.capacity = 1;
+  tx_pack.num_loaded = 0;
+  tx_pack.insufficient_capacity = false;
+  struct _iota_transaction curr_tx_s;
+  struct _iota_transaction next_tx_s;
+  iota_transaction_t curr_tx = &curr_tx_s;
+  iota_transaction_t next_tx = &next_tx_s;
+  tx_pack.models = (void **)(&curr_tx);
+  trit_array_t curr_tx_hash;
+  flex_trit_t bundle_hash[FLEX_TRIT_SIZE_243];
+  bool found_approver = false;
+  *found_tail = false;
+
+  res = iota_tangle_transaction_load(exit_probability_randomizer->tangle,
+                                     TRANSACTION_COL_HASH, tx_hash, &tx_pack);
+
+  if (res != RC_OK || tx_pack.num_loaded == 0) {
+    return res;
+  }
+
+  curr_tx = (iota_transaction_t)(tx_pack.models[0]);
+
+  uint32_t index = curr_tx->current_index;
+  memcpy(bundle_hash, curr_tx->bundle, FLEX_TRIT_SIZE_243);
+
+  iota_stor_pack_t hash_pack;
+  if ((res = hash_pack_init(&hash_pack, 10)) != RC_OK) {
+    return res;
+  }
+
+  while (res == RC_OK && index > 0 &&
+         memcmp(curr_tx->bundle, bundle_hash, FLEX_TRIT_SIZE_243) == 0) {
+    hash_pack.num_loaded = 0;
+    hash_pack.insufficient_capacity = false;
+    curr_tx_hash.num_bytes = FLEX_TRIT_SIZE_243;
+    curr_tx_hash.trits = curr_tx->hash;
+    res = iota_tangle_transaction_load_hashes_of_approvers(
+        exit_probability_randomizer->tangle, &curr_tx_hash, &hash_pack);
+
+    if (res != RC_OK) {
+      log_error(CW_RATING_CALCULATOR_LOGGER_ID,
+                "Failed in loading approvers, error code is: %\" PRIu64 \"\n",
+                res);
+      break;
+    }
+
+    --index;
+    uint32_t approver_idx = 0;
+    found_approver = false;
+    while (approver_idx < hash_pack.num_loaded) {
+      trit_array_p approver_hash =
+          (trit_array_t *)hash_pack.models[approver_idx];
+      tx_pack.capacity = 1;
+      tx_pack.num_loaded = 0;
+      tx_pack.insufficient_capacity = false;
+      tx_pack.models = (void **)(&next_tx);
+      res = iota_tangle_transaction_load(exit_probability_randomizer->tangle,
+                                         TRANSACTION_COL_HASH,
+                                         approver_hash, &tx_pack);
+      if (res != RC_OK || tx_pack.num_loaded == 0) {
+        break;
+      }
+      if (next_tx->current_index == index &&
+          memcmp(next_tx->hash, bundle_hash, FLEX_TRIT_SIZE_243) == 0) {
+        curr_tx = next_tx;
+        found_approver = true;
+        break;
+      }
+    }
+
+    if (!found_approver) {
+      break;
+    }
+  }
```

Comment:
Maybe we could write an SQL statement that would do that in a more efficient way ?
Why not to query txs by bundle hash ?